File: scala/55.scala
object Solution {
    def canJump(nums: Array[Int]): Boolean = {
        (0 until nums.length).foldLeft(0) {
            case (right, i) => {
                if (right >= nums.length - 1 || nums(i) == 0 || i > right) right
                else math.max(right, nums(i) + i)
            }
        } >= nums.length - 1
    }
}

File: leetcodeJava/55.java
package leetcodeJava;

class Solution {
    public boolean canJump(int[] nums) {
        var dp = new boolean[nums.length];
        dp[0] = true;
        for (var i = 0; i < nums.length; i++) {
            if (dp[i] == false) continue;
            for (var j = 1; j <= nums[i]; j++) {
                if (i+j >= dp.length)
                    return true;
                else
                    dp[i+j] = (dp[i] || dp[i+j]);
            }
        }
        return dp[dp.length-1];
    }
}